File: LinkedList/160. Intersection of Two Linked Lists/getIntersectionNode.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA == null || headB == null){
            return null;
        }
        //retrieve the tail and size of two list
        Result resultA = getTailandSize(headA);
        Result resultB = getTailandSize(headB);
        //No intersection if tails are not the same
        if(resultA.tail != resultB.tail){
            return null;
        }
        ListNode shorter = resultA.size < resultB.size ? headA : headB;
        ListNode longer = resultA.size < resultB.size ? headB : headA;
        //the longer list step forward the length difference
        int k = Math.abs(resultA.size - resultB.size);
        while(k > 0 && longer != null){
            longer = longer.next;
            k--;
        }
        //search the intersection point 
        while(shorter != longer){
            shorter = shorter.next;
            longer = longer.next;
        }
        return longer;        
    }
    class Result{
        public ListNode tail;
        public int size;
        public Result(ListNode tail, int size){
            this.tail = tail;
            this.size = size;
        }
    }
    Result getTailandSize(ListNode list){
        if(list == null){
            return null;
        }
        int size = 1;
        ListNode cur = list;
        while(cur.next != null){
            size++;
            cur = cur.next;            
        }
        return new Result(cur, size);
    }
}

File: LinkedList/160. Intersection of Two Linked Lists/getIntersectionNode.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    if(headA == NULL || headB == NULL){
        return NULL;
    }
    struct ListNode *head = headA;
    while(head->next != NULL){
        head = head->next;
    }
    //combine headB after the tail of headA
    //thus, we generate a cycle and the intersetion of two list is the node where the cycle begins.
    head->next = headB;
    //store the pointer of tail 
    struct ListNode *tail = head;
    //reset the pointer back to head
    head = headA;
    
    //fast and slow, determine if there is a cycle
    headB = headA;
    while(headB->next != NULL && headB->next->next != NULL){
        headA = headA->next;
        headB = headB->next->next;
        if(headA == headB){
            break;
        }
    }
    //no cycle
    if(headA->next == NULL || headB->next == NULL || headB->next->next == NULL){
        //break the chain 
        tail->next = NULL;
        return NULL;
    }
    //has cycle
    headA = head;
    while(headA != headB){
        headA = headA->next;
        headB = headB->next;
    }
    //break the chain
    tail->next = NULL;
    return headB;
}
